loadbalancer/algorithm.go

```diff
@@ -16,37 +19,83 @@ type algorithmType int
 const (
 	none algorithmType = iota
 	roundRobinAlgorithm
+	randomAlgorithm
+	consistentHashAlgorithm
 )
 
+var (
+	algorithms = map[algorithmType]initializeAgorithm{
+		roundRobinAlgorithm:     newRoundRobin,
+		randomAlgorithm:         newRandom,
+		consistentHashAlgorithm: newConsistentHash,
+	}
+	defaultAlgorithm = newRoundRobin
+)
+
+func newRoundRobin(endpoints []string) routing.LBAlgorithm {
+	i := time.Now().UnixNano()
+	rand.Seed(i)
+	return &roundRobin{
+		index: rand.Intn(len(endpoints)),
+	}
+}
+
 type roundRobin struct {
 	mx    sync.Mutex
 	index int
 }
 
-type algorithmProvider struct{}
+// Apply implements routing.LBAlgorithm with a roundrobin algorithm.
+func (r *roundRobin) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+	r.index = (r.index + 1) % len(ctx.Route.LBEndpoints)
+	return ctx.Route.LBEndpoints[r.index]
+}
+
+type random struct {
+	rand *rand.Rand
+}
 
-type initializeAgorithm func(endpoints []string) routing.LBAlgorithm
+func newRandom(endpoints []string) routing.LBAlgorithm {
+	t := time.Now().UnixNano()
+	return &random{rand: rand.New(rand.NewSource(t))}
+}
 
-var algorithms = map[algorithmType]initializeAgorithm{
-	roundRobinAlgorithm: newRoundRobin,
+// Apply implements routing.LBAlgorithm with a stateless random algorithm.
+func (r *random) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	return ctx.Route.LBEndpoints[r.rand.Intn(len(ctx.Route.LBEndpoints))]
 }
 
-var defaultAlgorithm = newRoundRobin
+type consistentHash struct{}
 
-func newRoundRobin(endpoints []string) routing.LBAlgorithm {
-	return &roundRobin{
-		index: rand.Intn(len(endpoints)),
-	}
+func newConsistentHash(endpoints []string) routing.LBAlgorithm {
+	return &consistentHash{}
 }
 
-// Apply implements routing.LBAlgorithm.
-func (r *roundRobin) Apply(endpoints []routing.LBEndpoint) routing.LBEndpoint {
-	r.mx.Lock()
-	defer r.mx.Unlock()
-	r.index = (r.index + 1) % len(endpoints)
-	return endpoints[r.index]
+// Apply implements routing.LBAlgorithm with a consistent hash algorithm.
+func (*consistentHash) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	var sum uint32
+	h := fnv.New32()
+
+	key := net.RemoteHost(ctx.Request).String()
+	if _, err := h.Write([]byte(key)); err != nil {
+		log.Errorf("Failed to write '%s' into hash: %v", key, err)
+		return ctx.Route.LBEndpoints[0]
```

Comment:
since fnv..Write never returns an error, this is just some nitpicking: if there was a chance for an error, then the fallback should be rather random instead of the first one